### ISSUE TITLE
Add defensive ndim==2 guard in preprocess_olive_weights to reject Olive moe=True 3D expert checkpoints

### ISSUE THREAD

**Issue by titaiwangms** [onnxruntime/mobius#427]
## Summary

`preprocess_olive_weights` in `src/mobius/_weight_utils.py` assumes every Olive-quantized weight tensor is 2-D (Linear / Embedding). Once Olive's upcoming `moe=True` RTN quantization lands, it will emit **3-D fused-expert** weight tensors `(num_experts, out_features, in_features)`. The current code has no `ndim`/shape guard, so it will silently mis-reshape those 3-D tensors and produce garbage weights with no error.

This issue asks only for a **defensive guard** that rejects (or explicitly refuses) 3-D expert tensors. Implementing actual MoE quantization support in Mobius is **out of scope** and tracked separately.

## Trigger

microsoft/Olive PR [microsoft/Olive#2584](https://github.com/microsoft/Olive/pull/2584) (resolving [microsoft/Olive#2583](https://github.com/microsoft/Olive/issues/2583)) adds a `moe=True` flag to Olive's PyTorch RTN quantization pass. With that flag, MoE expert weights are stored as **3-D fused tensors** `(num_experts, out_features, in_features)` in Olive's `QuantTensor` format, using the same buffer naming Mobius already keys on: `<pname>_qweight` / `<pname>_scales` / `<pname>_qzeros`.

Mobius already consumes Olive's **non-MoE** (2-D) checkpoints through `preprocess_olive_weights`, called from `src/mobius/models/base.py` (~line 451). Nothing in that path distinguishes 2-D from 3-D tensors.

## Where the guard is missing

`src/mobius/_weight_utils.py`, in the generic `.qweight` branch:

```python
727:        elif key.endswith(".qweight"):
728:            if value.dtype != torch.uint8:
729:                raise ValueError(f"Olive qweight must be uint8 for {key}, got {value.dtype}")
730:            if value.shape[-1] % blob_size != 0:
731:                raise ValueError(...)
732:            new_key = key.replace(".qweight", ".weight")
736:            result[new_key] = value.reshape(value.shape[0], -1, blob_size).contiguous()
```

**Line 736** is the silent-corruption site. For a 3-D expert tensor `(num_experts, out_features, packed_in)`:

- `value.shape[0]` is `num_experts` (not `out_features`), and
- `value.reshape(value.shape[0], -1, blob_size)` collapses `out_features * packed_in` into the middle axis,

producing a structurally valid but **semantically garbage** tensor — no exception is raised.

The sibling branches share the same blind assumption and would also need to be covered by whatever guard is added:

- embedding qweight branch — line 707 (`embed_tokens.qweight`)
- embedding qzeros branch — line ~721 (`embed_tokens.qzeros`)
- generic qzeros branch — line 737 (`.qzeros`)

Since `scales` / `qzeros` for a 3-D `qweight` would likewise be 3-D, the cleanest single choke point is a `value.ndim == 2` (or `> 2`) check applied to the Olive quantized buffers.

## Proposed fix (minimal, in-scope)

Add an explicit dimensionality guard so a 3-D Olive tensor produces a **clear, immediate error** instead of silent corruption. For example, in the `.qweight` (and, ideally, `.qzeros` / `embed_tokens.*`) branches:

```python
if value.ndim != 2:
    raise NotImplementedError(
        f"preprocess_olive_weights received a {value.ndim}-D tensor for {key}; "
        f"Olive moe=True 3D fused-expert checkpoints "
        f"(num_experts, out_features, in_features) are not yet supported by Mobius. "
        f"See microsoft/Olive#2583 / microsoft/Olive#2584."
    )
```

Anything equivalent (`assert value.ndim == 2`, an early check over the whole state dict, etc.) is acceptable — the requirement is simply **fail loudly on 3-D, never reshape it as 2-D**.

## Explicitly out of scope

- Implementing MoE quantization / 3-D fused-expert dequant in Mobius. Mobius's current MoE support is a plain `nn.ModuleList`-per-expert loop (`src/mobius/components/_moe.py`) with **zero** quantization wiring, so real support is a separate, much larger effort. This issue is purely a defensive guard against future silent corruption.

## References

- Olive issue: https://github.com/microsoft/Olive/issues/2583
- Olive PR: https://github.com/microsoft/Olive/pull/2584
- Affected code: `src/mobius/_weight_utils.py:736` (branch starting at line 727)
- Call site: `src/mobius/models/base.py` (~line 451)

**Comment by titaiwangms** [comment]
## Update: confirmed a live suffix-mismatch bug (blocks even today's 2-D Olive path), plus a full design doc for future MoE consumption — parked for now

While researching this issue further we had `local-architect` do a deeper pass on the current mobius loader (`_weight_utils.py`) vs. Olive's actual on-disk output as of `microsoft/Olive@d5c5641c`. Two findings worth recording here before we park this:

### 1. A suffix-mismatch bug, independent of MoE, that breaks the *existing* 2-D Olive path today

Olive's checkpoint key suffix convention is `<pname>_qweight` / `_scales` / `_qzeros` (underscore, e.g. `...gate_proj.weight_qweight` — see `olive/common/quant/state_dict.py:44-51`, `Olive test/common/quant/test_hf_utils.py:387`). The loader branches quoted in this issue's description match the **dotted** form (`.qweight`, `.scales`, `.qzeros`). As written, **any current Olive-quantized checkpoint (2-D or 3-D) fails to match these branches at all** and falls through silently rather than loading correctly. This is separate from (and more urgent than) the 3-D guard requested above — worth fixing regardless of whether/when MoE support lands, since it currently breaks plain `nn.Linear`/`nn.Embedding` quantized checkpoints too.

### 2. Full design doc for eventual MoE consumption (`QMoE` / `MatMulNBits` loop) — NOT being pursued right now

We had the architect draft a full design (loader + fused `QMoE` emission + per-expert `MatMulNBits` fallback), covering: on-disk layout contract, reuse of `models/gemma4.py`'s existing float fused-MoE emission as the closest precedent (rather than a new `integrations/olive/` package), an open orientation risk for gpt-oss-style N-last expert tensors (Olive currently quantizes blindly along the last dim with no check that it's actually K), the fact that Olive doesn't persist a `swiglu_fusion` discriminator on disk, and a proposed slice-by-slice sequencing (suffix fix → per-expert `MatMulNBits` loop on one model class → fused `QMoE` → follow-ups).

**We're explicitly not pursuing this right now** — the current priority for `microsoft/Olive#2584` is just producing a correct **PyTorch/HF-level** quantized checkpoint (torch semantics, no ONNX export attempted by Olive), not wiring up Mobius/ONNX consumption. Recording the design here so it isn't lost when this becomes a priority again. Full doc available on request (not yet committed anywhere permanent — currently a local artifact from this investigation).

### Recommendation for this issue's immediate scope

Given (1) above, consider widening this issue's minimal fix to also cover the suffix mismatch (both `_qweight`/`.qweight` conventions), not just the 3-D guard — otherwise the defensive guard proposed here will never actually trigger, since the 3-D tensors won't match the loader's key branches in the first place.

**Comment by justinchuby** [comment]
Maybe related: https://github.com/onnxruntime/mobius/pull/451
